src/utils/getFilteredEntityFields.ts

```diff
@@ -0,0 +1,181 @@
+import { useEntityFields, YextSchemaField } from "../hooks/index.ts";
+
+type Only<T, U> = {
+  [P in keyof T]: T[P];
+} & {
+  [P in keyof U]?: never;
+};
+type Either<T, U> = Only<T, U> | Only<U, T>;
+
+type ConfigFields<T extends Record<string, any>> = T["stream"]["fields"];
+
+type AllowList<T extends Record<string, any>> = {
+  allowList: ConfigFields<T>[number][];
+};
+
+type DisallowList<T extends Record<string, any>> = {
+  disallowList: ConfigFields<T>[number][];
+};
+
+type EntityFieldTypesFilter = {
+  types: EntityFieldTypes[];
+};
+
+export type RenderEntityFieldFilter<T extends Record<string, any>> = Either<
+  EntityFieldTypesFilter,
+  Either<AllowList<T>, DisallowList<T>>
+>;
+
+type EntityFieldTypes = "type.string" | "type.image" | `c_${string}`;
+
+const DEFAULT_DISALLOWED_ENTITY_FIELDS = [
+  "uid",
+  "meta",
+  "slug",
+  "c_visualConfigurations",
+  "c_pages_layouts",
+];
+const TOP_LEVEL_ONLY_FIELD_TYPES = ["type.image", "type.hours", "type.address"];
+
+type EntityFieldNameAndSchema = {
+  name: string;
+  schemaField: YextSchemaField;
+};
+
+/**
+ * Returns a list of {@link EntityFieldNameAndSchema} given a {@link YextSchemaField}.
+ * @param schemaField The schema field for which to get the field names.
+ * @returns Entity field names and their corresponding schema fields.
+ */
+const getEntityFieldNames = (
```

Comment:
We should have this take in the `filter { subfields }` so that we can use it wherever needed. Then getEntityTypeToFieldNames can be simplified. 

I'm wondering if we should also simplify or merge getFilteredEntityFields with this.

getFilteredEntityFields does:

1. getEntityFields (top level only)
2. filters disallowed
3. filters by allow/disallow list
4. gets all fields by type
5. which calls getEntityFieldNames for each top level field
6. gets the schema for each of those fields, makes a map
7. filter to the types

I realize I wrote this but I'll admit it's a little confusing. After talking with Aaron the allow/deny list will remain top level, so steps 1-3 are fine as is. But how could we easily return filteredEntityFields that includes the subfields? Most of that work already lives in getEntityFieldNames just without filtering. 

Btw, filtering by type would work at both the top and sub-levels while allow/deny is top-level only.

See what you can come up with. Another option could be to have getEntityFieldNames return more info about each field that indicates whether it's top level or not so that we could call that for step 1 and filter for steps 2-3. But it also feels weird to call getEntityFieldNames again within getEntityTypeToFieldNames.

Reply:
Updated so that it should handle subfields properly